### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] Nested mapped return types in curried functions false positive

### ISSUE THREAD

**Issue by eranhirsch** [typescript-eslint/typescript-eslint#9889]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.4&fileType=.tsx&code=C4TwDgpgBAyglgGwSAPAFQHxQLxQN5QDaAClHAHZQDWEIA9gGZRoC6AXMyS1AL4DcAWABQw0JFiJkAMTp10AGigA1AIYIArhCy54SVARJlKNek1YdVG6DwyCRQsdBly0iy5u34ipCtVqNmdmU1TV47YQATCABjBBUAJ2gGdXJo4Dg6SgZZBSgAYUyAZ2AVcmAMAAoIlRKOVyhojgLyYtLgAEoOZ1zm1rLbSJi4xKhk1PTM0Zz63pL%2Bisb8ormOjiqalTr2nCxumeW2gfso2ISklLSMykLJED3FWcP12uZFRceyzok9e6WWlaOJ2G53GVygNx%2B0weB3m7xhqygz02zG22CwumkUL%2BfXKdiAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Silly<T> = { [P in keyof T]: T[P] };

type SillyFoo<T, Value> = Silly<{ [P in keyof T]: Value }>;

type Foo<T, Value> = { [P in keyof T]: Value };

declare function foo<T, Constant>(data: T, c: Constant): Foo<T, Constant>;
declare function foo<T, Constant>(c: Constant): (data: T) => Foo<T, Constant>;

declare function sillyFoo<T, Constant>(data: T, c: Constant): SillyFoo<T, Constant>;
declare function sillyFoo<T, Constant>(c: Constant): (data: T) => SillyFoo<T, Constant>;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-parameters": "error",
  },
};
```

### Expected Result

Mapping the type doesn't change the semantics of the type parameter, it's used just the same.

### Actual Result

The rule fails to see "inside" the nested/recursive mapping.

### Additional Info

This happens in our codebase when a type is wrapped with type-fest's `Simplify` type. It looks like the double mapping is causing it.

This was surfaced when migrating Remeda to v8 (https://github.com/remeda/remeda/pull/801); it happens in this file, for example https://github.com/remeda/remeda/blob/main/src/mapValues.ts.

**Comment by bradzacher** [comment]
THis looks like a duplicate of #9721 which was supposed to have been fixed.
Or a duplicate of #9667 which similarly was supposed to have been fixed.

Must be some intersection of the two 🤦

**Comment by eranhirsch** [comment]
I can say for sure that #9721 was shipped, it reduced a lot of our bugs. This is one of the remaining cases after it was shipped in 8.3

**Comment by JoshuaKGoldberg** [comment]
[A bit simpler of a repro](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.tsx&code=C4TwDgpgBAkgJhAdsAlqAPAFQHxQLxSYDcAsAFDmiRQCyAhmJHFgDRQBqdANgK4S4F4SVBnJQoAbygBtAApQUiKAGsIIAPYAzQgF0AXB258oAX3LZSFMggDGXOgCdomnohup1SgM4ouXEABi6uroACJ0wHRsnLz8ABQ2BjF8AJQGcXARdAbhkSn4uPSMEMy5UYaxFkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false):

```ts
type Identity<T> = T;

type Mapped<T, Value> = Identity<
  { [P in keyof T]: Value }
>;

declare function sillyFoo<Data, Value>(c: Value): (data: Data) => Mapped<Data, Value>;
//                              ~~~~~
```

I tried debugging through this and got confused. The existing [`isTypeParameterRepeatedInAst`](https://github.com/typescript-eslint/typescript-eslint/blob/94f7c99550aa06df6d6a039d53494aeaf8356f7c/packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts#L103) function _should_ understand `Value` to be used twice: once in `c: Value` and once in `Mapped<Data, Value>`. The playground link shows the scope for it (`scopes` > `[FunctionScope$50](https://typescript-eslint.io/play/#FunctionScope$50)` > `variables` > `Variable#Value$1360`) having 2 references. But with a breakpoint in `isTypeParameterRepeatedInAST`, `Value` only has the 1 reference? Am I misreading something?

**Comment by kirkwaiblinger** [comment]
It looks like there is a references property on the `scope` itself, as well as each of the `scope.variables`. I think we might want the latter here? 

I tried something like this, and I do get the expected numbers of references, though it didn't solve the bug:
```diff
       const checker = parserServices.program.getTypeChecker();
       let counts: Map<ts.Identifier, number> | undefined;
 
+      const scope = context.sourceCode.getScope(node);
+
       for (const typeParameter of tsNode.typeParameters) {
         const esTypeParameter =
           parserServices.tsNodeToESTreeNodeMap.get<TSESTree.TSTypeParameter>(
             typeParameter,
           );
-        const scope = context.sourceCode.getScope(esTypeParameter);
+        const references = scope.variables.find(
+          variable =>
+            variable.identifiers.length === 1 &&
+            variable.identifiers[0] === esTypeParameter.name,
+        )!.references;
 
         // Quick path: if the type parameter is used multiple times in the AST,
         // we don't need to dip into types to know it's repeated.
         if (
           isTypeParameterRepeatedInAST(
             esTypeParameter,
-            scope.references,
+            references,
             node.body?.range[0] ?? node.returnType?.range[1],
           )
         ) {
```

**Comment by kirkwaiblinger** [comment]
See also 
- [docs for `references` property on `Scope`](https://eslint.org/docs/latest/extend/scope-manager-interface#references)
- [docs for `references` property on `Variable`](https://eslint.org/docs/latest/extend/scope-manager-interface#references-1)

**Comment by kirkwaiblinger** [comment]
> though it didn't solve the bug

🤦‍♂️ Yes, it did. I just put the case from the issue as an invalid case instead of valid 😆
